### DISCUSSION TITLE
Strategy for exporting from iCloud and preserving unique IDs of items

### DISCUSSION THREAD

**Discussion by hyfen** [RhetTbull/osxphotos#1516] [Q&A] [answered]
I'm trying to set up a workflow where I use osxphotos to periodically download photos and then import them into my main Lightroom catalog. 

Once in Lightroom, I'd love to have the option of querying details about a given photo in osxphotos again. Is there any metadata strategy I can use to make this easy? I was originally going to try and write the UUID to a metadata tag but am now seeing that they're only [unique to a local instance](https://github.com/RhetTbull/osxphotos/discussions/1386) so that won't survive in the long run.

Is the best I can do using the filename/time and querying osxphotos with that?

**Comment by hyfen** [comment]
Ah, I'm just realizing that the discussion on [this issue](https://github.com/RhetTbull/osxphotos/issues/1508#issuecomment-2053675620) is similar to what I was going to attempt to do but it still doesn't solve the issue of how to have a stable ID if you start using a new macos install right?

**Comment by RhetTbull** [comment, marked as answer]
UUID works as long as you stay on a single Mac but as you noted, if you try to compare to a library from another Mac (or in some cases, if you have to use the Photos library repair tool), macOS changes the UUID.  If you are using iCloud, there is a `Cloud GUID` globally unique identifier. There's also a fingerprint (a hash of some sort) that is unique to a given asset (though videos do not contain a fingerprint). Sometimes the GUID and the fingerprint are the same. For a sample image in my library, these have the value: `AcoaTOYrVq2IUcpsJ0MyUoKkcZcB` -- as you can see it's a random-looking string of characters. This could be encoded in the exported image -- perhaps in the caption or as the DocumentID/ImageUniqueID...though this expects a UUID and the GUID doesn't match the expected format.  You could then have an OSXPhotos script that finds the matching photo in Photos and opens it.  OSXPhotos already provides access to these via `{photo.cloud_guid}` and `{photo.fingerprint}` templates.

For example, there's a `osxphotos show UUID` command that given a UUID, finds the matching photo in Photos and opens it.  It would be possible to make this command also look for a GUID and a fingerprint and open the photo...this would be fairly easy. As I type this, I realize this is actually a good idea and I'll open a new issue for this.

If you never modify the exported image, the show command could compute the fingerprint and find the matching photo (`osxphotos import` already does this to avoid importing duplicates).  But if you modify the image, you'd need to somehow encode the fingerprint or GUID.  One option would be to use extended attributes (for example, in the Finder comments -- which OSXPhotos can write to with `--xattr-template findercomment`) or in a unique OSXPhotos extended attribute. This would only work if the file was exported to a filesystem that supported extended attributes.

**Reply by RhetTbull** [reply]
One approach would be to use the caption (description) field to store this if you were using `--exiftool` on export: `--description-template "{descr?{descr}{newline}}GUID:{photo.cloud_guid}"` would append (after a new line) `GUID:AQt2kDJ5RHXjvZ7EqO7O91iAPSak` to the caption/description. This could then be read by a script which would then find the corresponding photo in Photos.

**Reply by hyfen** [reply]
> It would be possible to make this command also look for a GUID _and_ a fingerprint

Should this be "or"? Shouldn't a single GUID be sufficient for looking up photo in the db or does it also need a fingerprint?

The EXIF `ImageUniqueID` tag doesn't require the value to be a UUID so I think only doing `exiftool -m -overwrite_original -ImageUniqueID='{photo.cloud_guid}` will work for my purposes. Or is there another reason I should also set the XMP DocumentID? (A new Document ID uuid will get generated by Lightroom and saved to the jpeg when metadata is synced back to the file)

**Reply by RhetTbull** [reply]
GUID or fingerprint should work. You'll only have a GUID if it's an iCloud library.

**Reply by RhetTbull** [reply]
As long as whichever app you're using doesn't expect `ImageUniqueID` to look like a UUID then just using the `cloud_guid` is probably fine. I don't know if any apps preferentially look at `DocumentID` over `ImageUniqueID` -- you'll need to test  your workflow to see if it matters.

**Reply by hyfen** [reply]
I'm trying to do a full export of my icloud library and I want the option of querying for more metadata (via osxphotos) in the future for any given file. I'll probably do this with my own script and it's reassuring to know that having `cloud_guid` somewhere in the metadata is sufficient for matching the file to a record in the db.

Thanks, this was very helpful.

**Reply by RhetTbull** [reply]
> I'm trying to do a full export of my icloud library and I want the option of querying for more metadata (via osxphotos) in the future for any given file. 

All the metadata for a given file is also stored in the OSXPhotos export database and this can be programmatically accessed via Python to extract what you need or update the files. Additionally, you can use `osxphotos exportdb /path/to/.osxphotos_export.db --info /path/to/exported/file` and it will print all the metadata in JSON format for the specified asset.

**Reply by hyfen** [reply]
Perfect, forgot about that option. I opened the export db and confirmed that this works (cloud_guid value copied from exported file's `Image Unique ID` exif tag):

```sql
SELECT
	*
FROM
	photoinfo
WHERE
	json_extract(photoinfo, '$.cloud_guid') = 'AaHJiUQeQ8Oj/Ru54UbmCp9LvcXo';
```
